2019Labs/CustomSecurityContent/documentation/lab3_oval.adoc

```diff
@@ -0,0 +1,487 @@
+INFO - Script line_not_there.fail.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
+INFO - Script correct_value.pass.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
+INFO - Script wrong_value.fail.sh using profile xccdf_org.ssgproject.content_profile_ospp OK
+....
+
+The output tells you the following:
+
+* The rule with full ID `xccdf_org.ssgproject.content_rule_accounts_tmout` has beeen tested in the `OSPP` profile context.
+* There were 4 test scenarios, all of them passed.
+* More information is available in the respective log directory.
+
+Now when we have a reasonable certainty about our rules, we can improve the OVAL content.
+
+
+== OVAL optimization
+
+The OVAL test repeats itself a little bit - there are checks for the `/etc/profile` file, as well as for others `/etc/profile.d/*.sh` files, but the tests and respective objects are very similar.
+This makes editing tedious and prone to copy-paste errors.
+Luckily, `ComplianceAsCode` supports the http://jinja.pocoo.org/docs/2.10/[jinja2] macro language that can be used to introduce some templating, thus removing this duplicity.
+
+Let's analyze what is the difference:
+
+The test has a different name and comment, plus it uses different test objects:
+Compare the two following excerpts:
+
+----
+<ind:textfilecontent54_test check="all" check_existence="all_exist" comment="TMOUT in /etc/profile"        id="test_etc_profile_tmout" version="1">
+...
+
+
+<ind:textfilecontent54_test check="all" check_existence="all_exist" comment="TMOUT in /etc/profile.d/*.sh" id="test_etc_profiled_tmout" version="1">
+...
+----
+
+
+Luckily, the jinja2 language enables us to define http://jinja.pocoo.org/docs/2.10/templates/#macros[macros] that can help us to remove the duplication.
+We are going to define a local macro that accepts the filename comment and the test stem as arguments.
+
+Therefore, we remove both tests, and we add the new macro and it's new invocations.
+
+----
+
+  {{% macro test_tmout(test_stem, files) %}}
+  <ind:textfilecontent54_test check="all" check_existence="all_exist" comment="TMOUT in {{{ files }}}" id="test_{{{ test_stem }}}" version="1">
+    <ind:object object_ref="object_{{{ test_stem }}}" />
+    <ind:state state_ref="state_etc_profile_tmout" />
+  </ind:textfilecontent54_test>
+  {{% endmacro %}}
+
+  {{{ test_tmout(  test_stem="etc_profile_tmout", files="/etc/profile") }}}
+  {{{ test_tmout(  test_stem="etc_profiled_tmout", files="/etc/profile.d/*.sh") }}}
+----
+
+Note that the delimiters are different than the `jinja2` website shows - i.e. instead of `{% macro ... %}` on the website, we use the `{{% macro ... %}}` form and so on - there is alway one curly brace more than the website documentation shows.
+
+Anyway, did we do everything right?
+Time to find out!
+Let's rebuild the datasteam, and let's execute the test suite again - the result should be exactly the same!
```

Comment:
We are overusing the exclamation marks in this document. I think a '.' will be more appropriate there because the exclamation mark feels like shouting.